### DISCUSSION TITLE
Can / should valarray use #pragma loop(ivdep)

### DISCUSSION THREAD

**Discussion by AlexGuteniev** [microsoft/STL#1172] [General] [answered]
Standard mentions the intention for valarray to optimize by assuming no dependencies between elements.

[[valarray.syn]/2](http://eel.is/c++draft/valarray.syn#2): 
> The valarray array classes are defined to be free of certain forms of aliasing, thus allowing operations on these classes to be optimized.

[[valarray.access]/4](http://eel.is/c++draft/numarray#valarray.access-4): 
> The expression addressof(a[i]) != addressof(b[j]) evaluates to true for any two arrays a and b and for any size_t i and size_t j such that i < a.size() and j < b.size(). [Note: This property indicates an absence of aliasing and may be used to advantage by optimizing compilers. Compilers may take advantage of inlining, constant propagation, loop fusion, tracking of pointers obtained from operator new, and other techniques to generate efficient valarrays. -- end note]

However, this is not enough to perform optimizations:
 * valarray can contain used defined type elements, with dependencies between their values
 * valarray operator argument may refer to an element or multpiple elements of it

[[valarray.assign]/10](http://eel.is/c++draft/valarray.assign#10)
> The value of an element in the left-hand side of a valarray assignment operator does not depend on the value of another element in that left-hand side

This is also not enough to perform optmimizations, as argument may be a view of the same array, so assigned and source elements may overlap

[[valarray.cassign]/2](http://eel.is/c++draft/valarray.cassign#2)
> The value of an element in the left-hand side of a valarray compound assignment operator does not depend on the value of another element in that left hand side.

Looks enough however. For this case even self-operand does not introduce inter-element dependencies.

So, if this analysis is correct, `#pragma loop(ivdep)` can be applied to `valarray`, but only to compound assignments.
Should this be done? Should there be a verbose comment?

**Comment by cbezault** [comment]
I'm not keeping up with the valarray PRs but if this was answered elsewhere could you throw the answer in here and we can mark it as accepted?

**Reply by AlexGuteniev** [reply]
No, it wasn't discussed anywhere else

**Comment by AlexGuteniev** [comment, marked as answer]
I can aswer myself: not in general case, as the elements of valarray can be of arbitrary types, for which this pragma isn't safe.

```c++
std::valarray<std::string> v; // compiles
```
